### PR TITLE
Proposals for Collapse, CollapseAll, Expand APIs

### PR DESCRIPTION
**Pull request by sandy081**
Implements https://github.com/Microsoft/vscode/issues/55879

### CODE REVIEW

**Comment by jrieken** [comment]
@sandy081 You can just merge proposed api changes, we will discuss them in the API call, and lets have a PR when merging this to stable

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -1142,7 +1142,37 @@ declare module 'vscode' {
...
 	//#endregion
 
-	//#region Tree Item Label Highlights
+	//#region Tree View
+
+	export namespace window {
+		export function createTreeView<T>(viewId: string, options: { treeDataProvider: TreeDataProvider<T> }): TreeView2<T>;
+	}
+
+	export interface TreeView2<T> extends TreeView<T> {
```
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/vscode.proposed.d.ts]: I tried using the same name 

```suggestion
	export interface TreeView<T> extends TreeView<T> {
```

but it caused recursive compilation error. Do you have other suggestions?

File: src/vs/vscode.proposed.d.ts [hunk 2]
```diff
@@ -1142,7 +1142,37 @@ declare module 'vscode' {
...
+		 *
+		 * @param elementOrElements element(s) to collpase.
+		 * @param recursive Controls if elements have to be collapsed recursively or not.
+		 */
+		collapse(elementOrElements: T | T[], recursive?: boolean): Thenable<void>;
+
+		/**
+		 * Collapse all elements.
+		 */
+		collapseAll(): Thenable<void>;
```
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/vscode.proposed.d.ts]: Tried this but felt weird `collapse` does collapsing all.

File: src/vs/vscode.proposed.d.ts [hunk 3]
```diff
@@ -1142,7 +1142,37 @@ declare module 'vscode' {
...
+
+	export interface TreeView2<T> extends TreeView<T> {
+
+		/**
+		 * Collapses an element or several elements.
+		 *
+		 * @param elementOrElements element(s) to collpase.
+		 * @param recursive Controls if elements have to be collapsed recursively or not.
+		 */
+		collapse(elementOrElements: T | T[], recursive?: boolean): Thenable<void>;
```
[the remark this answers is not in the document]
Comment (sandy081) [reply src/vs/vscode.proposed.d.ts]: `expand` is async and to be consistent I made `collapse` also. In fact `collapse` is async in the current tree impl. Not returning `Thenable` might cause authors to think it sync and can lead to unexpected errors.

**Pull request closed without merging**